### DISCUSSION TITLE
[Risolto] HA Core in Docker e integrazione shell command

### DISCUSSION THREAD

**Discussion by AlainGH** [indomus/forum#2318] [HUB: Home Assistant]
Salve a tutti, grazie alle guide di InDomus da oltre un anno ormai avevo HA Core installato su venv in un Raspberry  3B+.
Dopo che gli ultimi aggiornamenti mensili non filavano più lisci come all'inizio ma mi costringevano sempre a risolvere qualche problema con librerie non aggiornate ( o addirittura troppo aggiornate! : o) e incompatibilità varie ho deciso di migrare a Docker.

Purtroppo dopo poco mi sono accorto che una automazione che usa l'integrazione shell command ha smesso di funzionare e nel log veniva riportato l'errore 127.
Questa è la porzione relativa nel configuration.yaml:
```
shell_command:

  set_duty_cycle_to_slider: 'gpio pwm 1 {{ states("input_number.pwm_duty_cycle") }}'

  set_pwm_port: 'gpio mode 1 pwm ; gpio pwm-ms ; gpio clock 1 2400 ; gpio pwmr 1600 ; gpio pwm 1 0'
```
Lo stesso comando gpio funziona regolarmente se lanciato da finestra terminale attraverso una connessione vnc.
Leggendo in giro ho capito che HA eseguito dentro il container non "vede" i file presenti sull'host se non una piccola porzione tra cui la cartella di configurazione, raggiungibile attraverso il path `/config`.
Trattandosi dunque di un singolo comando ho provato a copiarlo dentro la cartella di configurazione che sull'host è `/home/homeassistant/.homeassistant` e nel configuration yaml ho aggiunto il path `/config/ ` davanti al comando facendolo diventare `/config/gpio...`.
Alla fine però nulla è cambiato, neppure dando a gpio i permessi chmod u+s.

Ho anche visto che una possibile soluzione consisterebbe nel lanciare il comando attraverso una connessione ssh all'host, usando una chiave d'accesso ad hoc, ma onestamente il mio basso livello di competenza non mi ha finora consentito di percorrere questa strada.
Vi sarei molto grato se poteste darmi una mano a risolvere questo mio piccolo problema

**Comment by snakuzzo** [comment]
Non so cosa tu abbia bisogno di fare nello specifico, innanzi tutto ti consiglierei di dare un'occhiata a [QUESTO](https://github.com/thecode/ha-rpi_gpio) custom component.

Se non ti soddisfa e vuoi fare come hai detto, ovvero collegandoti all'host in ssh, fai ad esempio così...

Premetto che questa cosa puoi farla dall'host che dall'interno del container. Di seguito ti faccio vedere come fare dall'interno del container:

Accedi alla shell del container

```
$ docker exec -it homeassistant bash
```

Crea una directory dove memorizzare in maniera persistente la chiave SSH

```
bash-5.1# mkdir -p /config/.ssh/
```

Crei la chiave specificando esplicitamente di crearla sotto la directory appena creata:

es.
```
bash-5.1# ssh-keygen -t rsa
Generating public/private rsa key pair.
Enter file in which to save the key (/root/.ssh/id_rsa): /config/.ssh/id_rsa
Enter passphrase (empty for no passphrase): 
Enter same passphrase again: 
Your identification has been saved in /config/.ssh/id_rsa
Your public key has been saved in /config/.ssh/id_rsa.pub
The key fingerprint is:
SHA256:xxxxxxxxxxxxxxxxxxxR/szY root@homeassistant
The key's randomart image is:
...
...
```

Poi copi la chiave pubblica tra le authorized keys dell'host.
Puoi farlo aggiungendo in coda al file `/home/snakuzzo/.ssh/authorized_keys` il contenuto del file `.pub` appena creato oppure usando `ssh-copy-id` con la seguente sintassi:

```
bash-5.1# ssh-copy-id -i /config/.ssh/id_rsa.pub utentehost@iphost
```

A questo punto, sempre dalla shell del container, puoi provare a collegarti così (devi sempre specificare con `-i` dove recuperare la key perché quella directory non è quella di default (visto che la devi rendere persistente ad ogni ricreazione del container)..

```
bash-5.1# ssh -i /config/.ssh/id_rsa utentehost@iphost
```

**Comment by AlainGH** [comment]
Ciao grande snakuzzo, mille grazie per il tuo intervento.
Avevo già notato quel custom component ma, per quanto mi è parso di capire, si limita a gestire sensori e switch.
Io ho bisogno di accedere al comando gpio per pilotare un segnale pwm e modulare una resistenza.

Intanto, da buon cyber-impedito ho seguito le tue dettagliate istruzioni per collegarmi via ssh all'host ma mi sono già arenato:
immagino che quando scrivi di aggiungere il contenuto di `id_rsa.pub` in coda al file authorized_keys significhi che questo file da qualche parte esiste già ma io finora non l'ho trovato.

Lanciando invece il comando `ssh-copy-id` mi viene restituito il seguente messaggio
```
bash-5.1# ssh-copy-id -i /config/.ssh/id_rsa.pub pi@192.x.x.x
/usr/bin/ssh-copy-id: INFO: Source of key(s) to be installed: "/config/.ssh/id_rsa.pub"
mktemp: (null): No such file or directory
/usr/bin/ssh-copy-id: ERROR: failed to create required temporary directory under ~/.ssh
```

Sicuramente ho commesso qualche errore...

**Reply by snakuzzo** [reply]
Prova a creare tu la directory `.ssh` e poi riprova a copiare la chiave

```
mkdir -p ~/.ssh
```

**Reply by Scialla** [reply]
però, all'aggiornamento del container di HA, perderà la chiave ssh. dovrà quindi ripetere le operazioni ad ogni aggiornamento, giusto?

**Reply by snakuzzo** [reply]
no...quella dir gli serve solo temporaneamente (forse)...la chiave ssh la salva sotto `/config/.ssh` (leggi bene cosa ho scritto)

**Reply by Scialla** [reply]
giusto, non avevo fatto caso al path

**Comment by AlainGH** [comment]
Creando manualmente la directory `.ssh`  la procedura sembra venga portata a termine, anche se non senza errori
```
bash-5.1# mkdir -p ~/.ssh
bash-5.1# ssh-copy-id -i /config/.ssh/id_rsa.pub pi@192.168.0.76
/usr/bin/ssh-copy-id: INFO: Source of key(s) to be installed: "/config/.ssh/id_rsa.pub"
The authenticity of host '192.168.0.76 (192.168.0.76)' can't be established.
ED25519 key fingerprint is SHAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? y
Please type 'yes', 'no' or the fingerprint: SHAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
/usr/bin/ssh-copy-id: INFO: attempting to log in with the new key(s), to filter out any that are already installed
expr: warning: '^ERROR: ': using '^' as the first character
of a basic regular expression is not portable; it is ignored
/usr/bin/ssh-copy-id: INFO: 1 key(s) remain to be installed -- if you are prompted now it is to install the new keys
pi@192.168.0.76's password: 
Connection closed by 192.168.0.76 port 22
```
e, facendo una prova di collegamento, il fatto che alla fine mi esca il prompt dell'host dovrebbe significare che ha funzionato
```bash-5.1# ssh -i /config/.ssh/id_rsa pi@192.168.0.76
pi@192.168.0.76's password: 
Linux _hostname_ 5.10.103-v7+ #1529 SMP Tue Mar 8 12:21:37 GMT 2022 armv7l

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
You have mail.
Last login: Wed Jan 25 21:43:21 2023
pi@_hostname_:~ $
```
Peccato che, dopo aver editato la porzione di `configuration.yaml` riportata ad inizio tread, sostituendo ogni `gpio` con la stringa `ssh -i /config/ssh/id_rsa -o 'StrictHostKeyChecking=no' pi@192.168.0.76 gpio`, in HA ancora non funziona e nel log appare l'errore 255.

Proverò a rifare tutto da capo per poi tornare a riferire

**Reply by snakuzzo** [reply]
Hai fatto tutto correttamente, quell'errore all'inizio puoi ignorarlo.
L'unica cosa che dovresti sistemare è il comando finale

**Reply by snakuzzo** [reply]
Hai sbagliato la directory

Se hai fatto come ti ho scritto, non è `/config/ssh/id_rsa`, ma `/config/.ssh/id_rsa`

**Reply by snakuzzo** [reply]
Ecco la prova fatta dal mio container docker di Core...

```
bash-5.1# ssh -i /config/.ssh/id_rsa -o 'StrictHostKeyChecking=no' snakuzzo@192.168.1.10 lsb_release -a
Distributor ID:	Debian
Description:	Debian GNU/Linux bookworm/sid
Release:	n/a
Codename:	bookworm
```

**Reply by snakuzzo** [reply]
Il comando provalo prima dalla shell del container...e poi lo riporti come shell_command (e qui stai attento agli apici)

**Comment by AlainGH** [comment]
Si `.ssh`, il nome della directory l'ho scritto male qui sopra ma sul raspberry è corretto.
Faccio qualche prova e poi ti dico

**Comment by AlainGH** [comment]
Ok, alla fine ce l'ho fatta.
La sintassi corretta per shell command è
`alias: "ssh -i /config/.ssh/id_rsa -o 'StrictHostKeyChecking=accept-new' utenteHost@IPHost 'comando originario'"`
per cui ho anche dovuto riscrivere l'istruzione `states("input_number.pwm_duty_cycle")` nella forma `states.input_number.pwm_duty_cycle.state` per evitare che gli apici e i doppi apici  venissero interpretati dal sistema come "termine del comando".
Ringrazio ancora Diego per avermi guidato e fornito le giuste dritte per giungere alla soluzione del problema
